src/engine/Match_tainting_mode.ml

```diff
@@ -132,6 +132,30 @@ type propagator_match = {
 }
 (** Taint will flow from `from` to `to_` through the axiliary variable `id`. *)
 
+(* This Formula_tbl structure is used to create a "formula cache", which will
+   permit the sharing of matches resulting from common formulas that occur as
+   sources, sinks, sanitizers, or propagators.
+
+   In particular, because of hardcoded propagators, we expect to see lots of
+   sharing from Semgrep Pro Engine.
+*)
+module Formula_tbl = struct
+  include Hashtbl.Make (struct
+    type t = Rule.formula
+
+    let equal = AST_utils.with_structural_equal Rule.equal_formula
+    let hash = Rule.hash_formula
+  end)
+
+  let cached_find_opt formula_cache formula cont =
+    match find_opt formula_cache formula with
+    | None ->
+        let ranges, expls = cont () in
+        replace formula_cache formula (ranges, expls);
```

Comment:
You can use `add` here given that you _know_ `formula` is not in the cache. Or is there another reason for `replace` ?

Reply:
yeah thats fair, i think i was just being overly paranoid and figured better safe than sorry, but I can change it